taichi/platform/opengl/opengl_api.cpp

```diff
@@ -0,0 +1,291 @@
+#include "opengl_api.h"
+
+#ifdef TI_WITH_OPENGL
+
+#define GLEW_STATIC
+#include <GL/glew.h>
+#include <GLFW/glfw3.h>
+#endif
+
+TLANG_NAMESPACE_BEGIN
+namespace opengl {
+
+#ifdef TI_WITH_OPENGL
+void glapi_set_uniform(GLuint loc, float value)
+{
+  glUniform1f(loc, value);
+}
+
+static std::string add_line_markers(std::string x)
+{
+  std::string marker;
+  size_t pos = 0, npos;
+  int line = 0;
+  while (1) {
+    npos = x.find_first_of('\n', pos);
+    marker = fmt::format("{:3d} ", ++line);
+    if (npos == std::string::npos)
+      break;
+    x.insert(pos, marker);
+    pos = npos + 1 + marker.size();
+  }
+  return x;
+}
+
+struct GLShader
+{
+  GLuint id_;
+
+  GLShader(GLuint type = GL_COMPUTE_SHADER)
+  {
+    id_ = glCreateShader(type);
+  }
+
+  GLShader(std::string source, GLuint type = GL_COMPUTE_SHADER)
+    : GLShader(type)
+  {
+    this->compile(source);
+  }
+
+  ~GLShader()
+  {
+    glDeleteShader(id_);
+  }
+
+  GLShader &compile(const std::string &source)
+  {
+    // Why strcpy? See https://stackoverflow.com/questions/28527956/get-011-error-syntax-error-unexpected-end-when-trying-to-compile-shader
+    GLchar *source_cstr = new GLchar[source.size() + 1];
```

Comment:
Should we `delete source_cstr` somewhere? I would directly use an `std::vector<GLchar>` and let C++ manage memory.